### DISCUSSION TITLE
multiple external libraries in Docker

### DISCUSSION THREAD

**Discussion by delirious11** [immich-app/immich#6912] [Q&A] [answered]
Hello.

I have immich setup in synology nas docker environment.    Upon setup, I altered two files.  The docker-compose.yml and the .env file.   I followed the guide at immich to set up my external library and I have the .env pointed to /photos.  https://immich.app/docs/guides/external-library 

 immich scans the library and adds the photos as expected.  What I'd like to do is add another share file from the NAS as another external library instead of copying over all my photos into /photo.    How do you do this?    My thoughts was to add another external path in the .env file    EXTERNAL_PATH=<your-path-here>  where as the second external library would point to another share folder /photos2

Is this correct?    The only parameter that designated where the external file is in the .env file.   The server and microservices have  - ${EXTERNAL_PATH}:/usr/src/app/external as instructed in the guide.   I am guessing those stay the same since they point to the .env file to designate location of the path for the photos.   any advice appreciated to add multiple external locations.   This seems to be inclusive of Docker related installs

**Comment by aisbergde** [comment]
you can use multiple external libraries, for example I use

```
   volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${IMMICH_LOCATION}/micro:/usr/src/app/.reverse-geocoding-dump
      # added for external libraries
      - ${EXTERNAL_PATH}/photo:/usr/src/app/external/common/photo:ro
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/photo:ro
    env_file:
      - stack.env
```

or you can use multiple variables. here you see a (not used) `EXTERNAL_PATH_GERMO`

```
IMMICH_LOCATION=/volume1/docker/immich
UPLOAD_LOCATION=/volume1/immich/upload
IMMICH_VERSION=release
TZ=Europe/Berlin
DB_HOSTNAME=immich_postgres
DB_DATABASE_NAME=immich
DB_USERNAME=immichuser
DB_PASSWORD=xxxxxxx
REDIS_HOSTNAME=immich_redis
EXTERNAL_PATH=/volume1
EXTERNAL_PATH_GERMO=/volume1/homes/germo
USER=1026:100

```

**Reply by mich2k** [reply]
![immagine](https://github.com/immich-app/immich/assets/13397578/27bc9268-d738-461d-aec3-48983b08f078)


but once different external sources are defined, how can I pass down from the admin to the user multiple locations?

**Reply by aisbergde** [reply]
as admin per user define the path, which can be used:
![Screenshot_2024-02-11-20-42-24-947_com.android.chrome-edit.jpg](https://github.com/immich-app/immich/assets/9020187/17f6301d-773d-43a9-9efd-dfd0e5aa5643)

and then as user add external libraries and use folders inside the allowed path

![Screenshot_2024-02-11-20-43-45-480_com.android.chrome-edit.jpg](https://github.com/immich-app/immich/assets/9020187/d9512fce-5369-4ad7-9ac9-9b1c93cd84e0)

![Screenshot_2024-02-11-20-45-02-251_com.android.chrome.jpg](https://github.com/immich-app/immich/assets/9020187/d754d04d-26c8-4b52-9233-67fcd08ea591)

![Screenshot_2024-02-11-20-44-47-789_com.android.chrome.jpg](https://github.com/immich-app/immich/assets/9020187/8fbe06bf-6591-4b84-9080-d6cfef1ec95f)

**Reply by aisbergde** [reply]
In my case only one external library would be enough to include personal and common library. Because I changed this later.
But I could also use these kind of libraries

- usr/src/app/external/common
- usr/src/app/external/photo
- usr/src/app/external/user_a
- usr/src/app/external/user_b
- usr/src/app/external/what_ever_would_match

if the mapping in the composite file would match

**Reply by mich2k** [reply]
But in this way if I have

/external/commondata
/external/userdataA
/external/userdataB

From admin I should pass the whole /external and the user A can add as external the sensible data of user B?

**Reply by aisbergde** [reply]
In my understanding as admin you have to "open" potentialy used folders, and then user  A should add one set of libraries and user B should add another set of libraries.

But it looks like the the question is, how to prevent user A to use the external libraries of user B and vice versa. This is a good question. because immich users are not mapped with NAS users.

Maybe a feature request is required, to implement this way of control? I also could imagine to move the user setting for external libraries into the admin area. Anyway the library settings are very technical and it is a bit hard to await from my family members library management.

**Reply by aisbergde** [reply]
I added #7048

**Comment by delirious11** [comment]
I opened a ticket on Discord in the immich help desk.  According to their advice, a user can only have one external library from the NAS.  Your photos can be in many subdirectories, but only one main share folder.  If you want to add a different shared folder, you have to log in as a different user.  For me,  I'll have to consolidate all my photos into the /photos folder in order to achieve access to multiple sub folders under one user.

**Reply by aisbergde** [reply]
can you share the link to the Discord thread?

**Comment by aisbergde** [comment]
this is a wrong advice, you can have multiple external libraries. Maybe some misunderstanding?

**Reply by delirious11** [reply]
I will if you are working with a synology NAS.  Which is what my question is about.   Are you?  if not, it may not apply to your application.  My question was specific about adding multiple Shared files on that NAS

**Reply by aisbergde** [reply]
DS923+

**Reply by delirious11** [reply]
I have set up everything on my second nas as you have.   I don't get any movement once I scan my shared folder.  I have only one setup in the admin/library console for now, but that one isn't scanning anything

**Reply by aisbergde** [reply]
Do you use a specific user?
Which group ID for this user do you use?
for example one of these
- USER=1026:100
- USER=1026:101

did  you care about permissions?
In the case of group "100" ("users" on my DS) you need to grant explicit permissions, even if the user is in the admin group. If you use "101" ("administrators") the user should already have access.

Use a console inside your containers and check the content of the mounted folders, do you see the assets?

I used portainer to manage the docker containers. In this case you can easily use a console.

**Reply by delirious11** [reply]
I did use 1026:100 (verified when I putty into the nas and check ID). (should I use 101?)  Permissions were also set for access on the shared folders on the nas for admin, user and everyone for that matter.  

 My issue, and I tried all sort of scenarios you listed, is that the container wants a path.    I have two shared folders setup for a test.  /photo and /photo2.     In the container settings in the docker for the immich server and microservies, It wants me to map the volumes to /photo  (nas) and docker /usr/src/app/external.   I can see the photos in the immich app if I designate the administration library to /usr/src/app/external.  It shows the  photos in /photo( set in the docker settings for the server and microservices).  But I can't designate a second location in the docker settings.   see photo
![x](https://github.com/immich-app/immich/assets/63081228/a678cfe9-b181-4c15-bb2e-8bc1ea0cb101)  Only one path is accepted with /usr/src/app/external

I have to go to work soon.   Probably will try this after work this evening.   

Here are my settings in the yml and env

version: "3.8"



name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${EXTERNAL_PATH}volume1/photo:/usr/src/app/external/photo:ro
      - ${EXTERNAL_PATH}volume1/photo2:/usr/src/app/external/photo2:ro
      - /etc/localtime:/etc/localtime:ro
      
    env_file:
      - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: always

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/hardware-transcoding
    #   file: hwaccel.transcoding.yml 
    #   service: cpu # set to one of [nvenc, quicksync, rkmpp, vaapi, vaapi-wsl] for accelerated transcoding
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${EXTERNAL_PATH}volume1/photo:/usr/src/app/external/photo:ro
      - ${EXTERNAL_PATH}volume1/photo2:/usr/src/app/external/photo2:ro
      - /etc/localtime:/etc/localtime:ro
      
    env_file:
      - .env
    depends_on:
      - redis
      - database
    restart: always

  immich-machine-learning:
    container_name: immich_machine_learning
    # For hardware acceleration, add one of -[armnn, cuda, openvino] to the image tag.
    # Example tag: ${IMMICH_VERSION:-release}-cuda
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    # extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/ml-hardware-acceleration
    #   file: hwaccel.ml.yml
    #   service: cpu # set to one of [armnn, cuda, openvino, openvino-wsl] for accelerated inference - use the `-wsl` version for WSL2 where applicable
    volumes:
      - ./model-cache:/cache
    env_file:
      - .env
    restart: always

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:afb290a0a0d0b2bd7537b62ebff1eb84d045c757c1c31ca2ca48c79536c0de82
    restart: always

  database:
    container_name: immich_postgres
    image: tensorchord/pgvecto-rs:pg14-v0.1.11@sha256:0335a1a22f8c5dd1b697f14f079934f5152eaaa216c09b61e293be285491f8ee
    env_file:
      - .env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - ./pgdata:/var/lib/postgresql/data
    restart: always


and .........


UPLOAD_LOCATION=./library
EXTERNAL_PATH=/volume1
EXTERNAL_PATH_photo=/volume1/photo
EXTERNAL_PATH_photo2=/volume1/photo2

TYPESENSE_API_KEY=asdfasdraeraseraser
DB_PASSWORD=postgres

DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich

REDIS_HOSTNAME=immich_redis


some of this is probably wrong but I was trying all sorts of scenarios and then quit and went to bed.  I do have portainer and am learning how to use it, but not proficient.   Honestly, all this is new to me.   I have it working on my primary NAS pointed to a shared folder.  I just want to add another shared folder and that's where the problem lies

**Reply by aisbergde** [reply]
it is a bit hard to read your code, because it is shown as "normal" text.

- did you define the same Volumes in both containers `immich_server` and `immich_microservices`?
- did you set the user in both containers?

```
# https://docs.docker.com/compose/compose-file/compose-file-v3/
# https://forums.docker.com/t/docker-compose-version-3-8-or-3-9-for-latest/102439/2
# both work, 3.8 and 3.9
version: "3.9"

#
# WARNING: Make sure to use the docker-compose.yml of the current release:
#
# https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml
#
# The compose file on main may not be compatible with the latest release.
# needs to be adapted for Synology
#

name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      # added for external libraries
      - ${EXTERNAL_PATH}/photo:/usr/src/app/external/common/photo:ro
      # the next entry will be removed after migrating from /usr/src/app/external/photo to /usr/src/app/external/germo
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/photo:ro
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/germo:ro
      - ${EXTERNAL_PATH}/homes/felicia/Pictures:/usr/src/app/external/felicia:ro
    env_file:
      - stack.env
      # original did not work:
      # - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: always
    # user - added for synology
    # attention!  1026 is an administrator, but doesn't get administrator rights, but only rhights of the group 100 - users
    # could be added  as 1026:101 (not tested)
    # or for the user 1026 explicit rights to used folders needs to be granted
    # How to get userID, see 
    # https://www.synoforum.com/resources/how-to-find-uid-userid-and-gid-groupid.77/
    # https://mariushosting.com/synology-find-uid-userid-and-gid-groupid-in-5-seconds/
    user: ${USER}
    # added from here (https://mariushosting.com/how-to-install-immich-on-your-synology-nas/), not sure, if this is required
    security_opt:
      - no-new-privileges:true

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends:
    #   file: hwaccel.yml
    #   service: hwaccel
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${IMMICH_LOCATION}/micro:/usr/src/app/.reverse-geocoding-dump
      # added for external libraries
      - ${EXTERNAL_PATH}/photo:/usr/src/app/external/common/photo:ro
      # the next entry will be removed after migrating from /usr/src/app/external/photo to /usr/src/app/external/germo
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/photo:ro
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/germo:ro
      - ${EXTERNAL_PATH}/homes/felicia/Pictures:/usr/src/app/external/felicia:ro
    env_file:
      - stack.env
      # original doesn't work
      # - .env
    depends_on:
      - redis
      - database
    restart: always
    # user - added for synology
    user: ${USER}
    # added from here (https://mariushosting.com/how-to-install-immich-on-your-synology-nas/), not sure, if this is required
    security_opt:
      - no-new-privileges:true

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
#      - model-cache:/cache
      - ${IMMICH_LOCATION}/cache:/cache
    env_file:
      - stack.env
      # original doesn't work
      # - .env
    restart: always
    # user - added for synology
    user: ${USER}
    # added from here (https://mariushosting.com/how-to-install-immich-on-your-synology-nas/), not sure, if this is required
    security_opt:
      - no-new-privileges:true

  redis:
    container_name: immich_redis
    image: redis
    restart: always

  database:
    container_name: immich_postgres
    # image: tensorchord/pgvecto-rs:pg14-v0.1.11@sha256:0335a1a22f8c5dd1b697f14f079934f5152eaaa216c09b61e293be285491f8ee
    # original did not work, this is from https://mariushosting.com/how-to-install-immich-on-your-synology-nas/
    image: tensorchord/pgvecto-rs:pg16-v0.1.11
    env_file:
      - stack.env
      # original doesn't work
      # - .env
    environment:
      TZ: Europe/Berlin
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      # changed to make it somehow working
      - ${IMMICH_LOCATION}/db:/var/lib/postgresql/data
    restart: always
    # added for external access.
    # this did not work, port conflict.
    # 5432:5432
    # i tried somethind different to make it work
    ports:
      - 7432:5432
    # added from here (https://mariushosting.com/how-to-install-immich-on-your-synology-nas/), not sure, if this is required
    healthcheck:
      test: ["CMD", "pg_isready", "-q", "-d", "${DB_DATABASE_NAME}", "-U", "${DB_USERNAME}"]
      interval: 10s
      timeout: 5s
      retries: 5
    # added from here (https://mariushosting.com/how-to-install-immich-on-your-synology-nas/), not sure, if this is required
    security_opt:
      - no-new-privileges:true

  # added for automatic database dump
  # found somewhere in the immich docs
  backup:
    container_name: immich_db_dumper
    image: prodrigestivill/postgres-backup-local
    env_file:
      - stack.env
      # original doesn't work
      # - .env
    environment:
      POSTGRES_HOST: database
      POSTGRES_DB: ${DB_DATABASE_NAME}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      SCHEDULE: "@daily"
      BACKUP_DIR: /db_dumps
    volumes:
      - ${EXTERNAL_PATH}/homes/germo/Backup/immich/db_dumps:/db_dumps
    depends_on:
      - database

  adminer:
    container_name: immich_adminer
    image: adminer
    ports:
      - 8080:8080
    environment:
      ADMINER_DEFAULT_SERVER: immich_postgres
    depends_on:
      - database

# I have no idea, what should this be, it works fine without
#volumes:
#  pgdata:
#  model-cache:

```

administration > user management - repeat settings for each user:

![image](https://github.com/immich-app/immich/assets/9020187/11f9144f-255f-447e-a0f8-02cc11e99b10)

and then each users needs his own external libraries
![image](https://github.com/immich-app/immich/assets/9020187/3cb35558-9038-41dc-9ba0-7540ecfdf522)

optional add some exclusion patterns, if you use or used Synology photos or other Synology apps, adding some additional files
![image](https://github.com/immich-app/immich/assets/9020187/018669ed-d59f-4985-958c-f85b356b58b2)

**Reply by delirious11** [reply]
I did not add user for both containers.  will try this when I get home from work (along with a few things you listed).  I really appreciate the help you are given and it's not taken for granted.

**Comment by delirious11** [comment]
https://discord.com/channels/979116623879368755/1203772557078040659

BTW, the person who replied, wrote many of the articles at this website.   https://immich.app/docs/overview/introduction

**Reply by aisbergde** [reply, marked as answer]
it is a misunderstanding.
Alex wrote (and this is correct), that you can have only one "external path" per user. For example my one (1) and only external path is
`/usr/src/app/external`

But you can mount different NAS-folders into this virtual "external path", and this is what I tried to explain:
```
      - ${EXTERNAL_PATH}/photo:/usr/src/app/external/common/photo:ro
      - ${EXTERNAL_PATH}/homes/germo/Photos:/usr/src/app/external/photo:ro
```
both are mounted inside immich under `/usr/src/app/external`, but they are located on completely different places in the NAS file system

And now you can
- add the whole `/usr/src/app/external` as one external library
  - this could be wrong, because it could add assets of other users to my external library. But I did the setup, when I was the one and only user  
- or add different external libraries for subfolders of `/usr/src/app/external`
  - /usr/src/app/external/common/photo
  - /usr/src/app/external/photo

I can mount the next user as
```
      - ${EXTERNAL_PATH}/homes/felicia/Photos:/usr/src/app/external/felicia:ro
```

and then the Synology user "felicia", which has an associated immich user felicia, can add
- two libraries
  - /usr/src/app/external/common/photo
  - /usr/src/app/external/felicia

I am not sure, if the "common" library is a good idea for now, because photos could be duplicated
I stopped to use the common folder, because I am waiting for a solution to not duplicate assets in common libraries.

**Reply by delirious11** [reply]
I have two synology NAS.   Right now, I have things setup for immich to download photos from my iphone and that's primarily what I use it for.   I would like to add more external libraries and will try what you posted above.  Thank you for that.   I think I'll use the second NAS to run some experiments instead of loosing what I have on my primary NAS.  I use the second NAS to back up the first.   If all goes well, I'll add those external libraries on the primary.

**Reply by delirious11** [reply]
Got this working!  Thanks for all your help.  I was using an install from some other source.  I decided to go with what you posted and use Marius hosting's install and got everything setup.  Seeing both shared folders now.

**Reply by aisbergde** [reply]
I also started with Mario Hostings article. But then I decided to compare with the immich suggest compose file and adapted a mix, until it worked.

I think, it would be a good idea to add some synology specific comments to the immich suggested compose file to make life easier for Synology users.

Another idea would be to add a specific article to the immich docs, but my understanding was (it could be a misunderstanding) that this idea was rejected to not bloat the docs.

**Comment by skrilla0** [comment]
@aisbergde Hello, I don't mean to hijack this discussion, but I am having the same issue supporting two external libraries from separate folder locations on my NAS. I believe it's my formatting of the docker-compose file, but banging my head understanding the root case on why it's not working 

As of right now this is the docker compose file in both services and microservices which works for just one external library: ` /mnt/immich:/mnt/immich:ro`

```
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /mnt/immich:/mnt/immich:ro
      - /home/skrilla/icloud:/mnt/immich:ro
      - /etc/localtime:/etc/localtime:ro
```     

I tried following your formatting as:

```
      volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${EXTERNAL_PATH}/mnt/immich:/mnt/immich:ro
      - ${EXTERNAL_PATH}/home/skrilla/icloud:/mnt/immich:ro
```

When rebuilding the docker container, I'm getting the following error:

`WARN[0000] The "EXTERNAL_PATH" variable is not set. Defaulting to a blank string. `

My experience with Docker is limited. What am I missing?

**Reply by delirious11** [reply]
here's my code for server and micro services volumes: (both volumes should match)

volumes:
      - /volume1/docker/immich/upload:/usr/src/app/upload
      - ${EXTERNAL_PATH}/icloud:/usr/src/app/external/icloud:ro
      - ${EXTERNAL_PATH}/photo:/usr/src/app/external/photo:ro

the shared folders are case sensative.  Noticed if I put in Photo, it didn't work.  photo did.


env:

NODE_ENV=production
TZ=america/chicago
DB_HOSTNAME=immich-db
DB_USERNAME=immichuser
DB_PASSWORD=immichpw
DB_DATABASE_NAME=immich
REDIS_HOSTNAME=immich-redis
EXTERNAL_PATH=/volume1
LOG_LEVEL=log
JWT_SECRET=xia7jNMRT50PgcShC0BTuFxTLPamoYnQO8OKH8jUff4iD0AdtbC8HTZp/Rv4LSdpV7kXbYTvxNrw24NzJJiag2nisYbUmEV8XbJEEboag98z+6KR5djhPPC4/oztoSJHOB2rpPXjv4VNwCAT0PvTKUo3gSY8rDja85DGugYqJ0c=
DISABLE_REVERSE_GEOCODING=false
REVERSE_GEOCODING_PRECISION=3
IMMICH_WEB_URL=http://immich-web:3000
IMMICH_SERVER_URL=http://immich-server:3001
IMMICH_MACHINE_LEARNING_URL=http://immich-machine-learning:3003


Immich app is /usr/src/app/external for administration path


library external path is 

/usr/src/app/external/photo
/usr/src/app/external/icloud

I'd only load one external at a time so you don't overload the NAS with a large number of files.  Wait until one populated thumbnails, then add the other.  

also make sure your yml has appropriate user id.   example 1026:100.    Putty ssh into your Nas and see what your numbers are.

**Reply by aisbergde** [reply]
I changed the user from 1026:100 to 1026:101

100 - group users
101 - group administrators

If using 100 (users), then explicit permissions needs to be granted on every mounted path.

using 101 will use the already existing administrator rights.


Regarding the missing environment variable for the external path:
I followed the instruction on Mario Hosting to install portainer on my Synology NAS. And inside portainer it is easy to manage the docker containers.
